extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -331,12 +335,63 @@ export class AmpAnalytics extends AMP.BaseElement {
       request = addParamsToUrl(request, this.config_['extraUrlParams']);
     }
 
-    this.config_['vars']['requestCount']++;
 
+    return this.isSampledIn_(trigger).then(result => {
+      if (!result) {
+        return Promise.resolve('');
+      }
+
+      this.config_['vars']['requestCount']++;
+      request = this.expandTemplate_(request, trigger);
+
+      // For consistency with amp-pixel we also expand any url replacements.
+      return urlReplacementsFor(this.getWin()).expand(request).then(request => {
+        this.sendRequest_(request, trigger);
+        return request;
+      });
+    });
+  }
+
+  /**
+   * @param {!JSONObject} trigger The config to use to determine sampling.
+   * @return {!Promise.<boolean>} Whether the request should be sampled in or
+   * not based on sampleSpec.
+   * @private
+   */
+  isSampledIn_(trigger) {
+    const spec = trigger['sampleSpec'];
+    const resolve = Promise.resolve(true);
+    if (!spec) {
+      return resolve;
+    }
+    const threshold = spec['threshold'];
+    if (!spec['sampleOn'] ||
+        Number.isNaN(parseFloat(threshold)) || !Number.isFinite(threshold)) {
+      console./*OK*/error(this.getName_(), 'Invalid sampling spec.');
+      return resolve;
+    }
+    const key = this.expandTemplate_(spec['sampleOn'], trigger);
```

Comment:
So, we should definitely not hash anything on each actual event. That is what I meant. The user's sample state should only be calculated once. If this was e.g. used on an interval or active-use trigger it would get really expensive.